### PR TITLE
Project elevation properties framework

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR implements the framework for the project-level elevation settings described in https://github.com/qgis/QGIS-Enhancement-Proposals/issues/246

Elevation/terrain settings are now project-level settings, which are defined through the Project Properties, Terrain tab:

image

Options are available for:

- Flat terrain: with terrain height setting
- DEM (Raster Layer): with setting for raster layer, vertical scale and offset
- Mesh: with setting for mesh layer, vertical scale and offset

When any NEW 3d map is created in the project, the map will default to using the same terrain settings as are defined for the project. (This makes it easier to create and destroy 3d maps on a whim, without having to re-setup their terrain settings from scratch each time!). 

The project level elevation settings will also be respected by the incoming Profile tool.

### CODE REVIEW

**Comment by wonder-sk** [comment]
Thanks @nyalldawson - my comments on the code so far:
- the terrain providers are called "project terrain providers" and live in src/core/project - but it would also make sense for some 3rd party code to instantiate a terrain provider based on a layer and use it somehow to query elevations, without any link to a project. Maybe we can drop the "project" part from terrain providers naming, and they could live in a "neutral" location, e.g. src/core/elevation ?
- how would be terrain providers dealing with multi-threaded access? are we going to make copies of providers in places where they are needed by background jobs (3d views, profile views) or have a single instance and use locking?
- I am wondering how to connect these new terrain providers with the terrain generators in 3D library that take care of generating terrain mesh geometry, do you have any thoughts on that? (I agree with your design to have in core lib just the logic to extract the elevation values and leave all the logic for 3D geometry generation to 3D library)
- heightAt() - in 3D library, this function has been quite tricky to implement. For rasters we still have a primitive implementation that uses 1024x1024 coarse resolution of the raster to avoid lots of raster requests. So for big datasets (e.g. dem for the whole world) we are not doing terrain clamping well (and for the "online" terrain in 3D we skip it altogether). I am wondering how we can improve the heightAt() interface to provide correct results but also with reasonable I/O cost. It may be worth passing a whole vector of X,Y points for which we'll need elevation, so that the provider can batch any raster block requests. Also, it may be useful to give the provider a sense of accuracy we need (horizontal or vertical resolution?) - so that the provider knows whether the request for elevations can be fulfilled from lower resolution or if data should be fetched using highest resolution possible.
- in heightAt() call, I would suggest to use NaN when we're outside of the layer extent or the layer is not available (currently returning zero)

**Comment by nyalldawson** [comment]
@wonder-sk 

> the terrain providers are called "project terrain providers" and live in src/core/project - but it would also make sense for some 3rd party code to instantiate a terrain provider based on a layer and use it somehow to query elevations, without any link to a project. Maybe we can drop the "project" part from terrain providers naming, and they could live in a "neutral" location, e.g. src/core/elevation ?

Sounds reasonable, I'll do that.

> how would be terrain providers dealing with multi-threaded access? are we going to make copies of providers in places where they are needed by background jobs (3d views, profile views) or have a single instance and use locking?

Copies are what I'm thinking. I'll add a "clone" method to handle this.

> I am wondering how to connect these new terrain providers with the terrain generators in 3D library that take care of generating terrain mesh geometry, do you have any thoughts on that? (I agree with your design to have in core lib just the logic to extract the elevation values and leave all the logic for 3D geometry generation to 3D library)

I was thinking of just having some logic when a new 3d view is created to set the initial 3d map settings terrain settings to match as much as possible of the project elevation settings. So we keep these two completely separate, and just have some conversion code sitting at app level to convert from project settings -> 3d map terrain settings. (In other words there's no changes to the terrain generators themselves at all)

> heightAt() - in 3D library, this function has been quite tricky to implement. For rasters we still have a primitive implementation that uses 1024x1024 coarse resolution of the raster to avoid lots of raster requests. So for big datasets (e.g. dem for the whole world) we are not doing terrain clamping well (and for the "online" terrain in 3D we skip it altogether). I am wondering how we can improve the heightAt() interface to provide correct results but also with reasonable I/O cost. It may be worth passing a whole vector of X,Y points for which we'll need elevation, so that the provider can batch any raster block requests. Also, it may be useful to give the provider a sense of accuracy we need (horizontal or vertical resolution?) - so that the provider knows whether the request for elevations can be fulfilled from lower resolution or if data should be fetched using highest resolution possible.

Right now the DEM terrain just uses the "sample" method, which is not so efficient but should be pixel accurate. I am thinking this "heightAt" method is really just a crude implementation useful for when someone wants to query an individual point only.

I would then see a "prepare" method added at some stage, which takes a map settings or render context or just extent/scale, and does whatever useful things are necessary to prefetch content so that the provider is ready for bulk queries. Then e.g. the DEM provider would fetch the appropriate block upfront, and just do the queries directly on this. (Prepare would be called in the background thread on the cloned provider, so the preparation is only valid for one profile render).

> in heightAt() call, I would suggest to use NaN when we're outside of the layer extent or the layer is not available (currently returning zero)

Sounds good, i'll do

**Comment by wonder-sk** [comment]
> I was thinking of just having some logic when a new 3d view is created to set the initial 3d map settings terrain settings to match as much as possible of the project elevation settings. So we keep these two completely separate, and just have some conversion code sitting at app level to convert from project settings -> 3d map terrain settings. (In other words there's no changes to the terrain generators themselves at all)

Sounds good. I assume the terrain generators will just stop having their own heightAt() implementation and use the one for core terrain providers, right?

> Right now the DEM terrain just uses the "sample" method, which is not so efficient but should be pixel accurate. I am thinking this "heightAt" method is really just a crude implementation useful for when someone wants to query an individual point only.

With the "sample" method without a prepare() method I am just worried that this will be terribly slow, especially because often we need thousands of elevations when preparing 3D entities for buildings.

What about the resolution for heightAt/prepare? It feels like that could save us a lot of work if the accurate values are not strictly needed... E.g. when looking at a 3D entity from far away, the heights do not need to be exact.

**Comment by nyalldawson** [comment]
@wonder-sk I've just updated this PR so that it's now at what I consider a merge-ready state. (It's ready for re-review now)

In answer to the outstanding questions:

> Sounds good. I assume the terrain generators will just stop having their own heightAt() implementation and use the one for core terrain providers, right?

Possibly at some future stage, but for now I'm not touching the 3d terrain generators at all. All we do is ensure that the generator is initialised and setup to match the project's elevation settings whenever a new 3d map is created.

> With the "sample" method without a prepare() method I am just worried that this will be terribly slow, especially because often we need thousands of elevations when preparing 3D entities for buildings.

Yes, it likely will be, but (as per the note above) we won't be using this for the 3d entities at this stage. So I'm planning on just using the sample approach for the profile tool initially, and then come back and re-assess the prepare method when we've got the bulk of the profile tool implemented.

> What about the resolution for heightAt/prepare? It feels like that could save us a lot of work if the accurate values are not strictly needed... E.g. when looking at a 3D entity from far away, the heights do not need to be exact.

Again, there's no changes here for the 3d entities, so we can defer this consideration for now.

**Review by wonder-sk — APPROVED** [review]

**Comment by wonder-sk** [comment]
Thanks for your answers @nyalldawson - all that makes perfect sense.

Thinking about the terrain options a bit more, wondering if it would make sense to also have an option to have no terrain at all. 3D views now also support visualization with no terrain. I guess in some cases it would be better not to have terrain defined rather than pretending that there's flat terrain at elevation zero, for example, if I load a point cloud and want to do some profiles of it, the fake flat terrain of the project may get in the way. But I'm just thinking out loud here...

File: src/app/project/qgsprojectelevationsettingswidget.cpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  mDemScaleSpinBox->setClearValue( 1.0 );
+  mComboDemLayer->setFilters( QgsMapLayerProxyModel::RasterLayer );
+
+  mMeshOffsetSpinBox->setClearValue( 0.0 );
+  mMeshScaleSpinBox->setClearValue( 1.0 );
+  mComboMeshLayer->setFilters( QgsMapLayerProxyModel::MeshLayer );
+
+  mComboTerrainType->addItem( tr( "Flat Terrain" ), QStringLiteral( "flat" ) );
+  mComboTerrainType->addItem( tr( "DEM (Raster Layer)" ), QStringLiteral( "raster" ) );
+  mComboTerrainType->addItem( tr( "Mesh" ), QStringLiteral( "mesh" ) );
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/app/project/qgsprojectelevationsettingswidget.cpp:39]: Personally I think it'd be a small step back in usability (an empty layer= flat terrain is not so obvious to me), with also some cost of future flexibility. (Eg I suspect we may need a "band" choice at some stage for the raster terrain provider, and likely mesh provider will have some mesh specific options at some stage too.)

**Comment by nyalldawson** [comment]
@wonder-sk 

> the fake flat terrain of the project may get in the way.

To me a "no terrain" option would be exactly identical to the default flat terrain which has zero offset. In both cases we are just taking heights/z values direct from a layer and leaving them unchanged. Unless I'm missing something here? 

(I'll merge this to keep things flowing, but am willing to revisit this point if I'm missing something!)

**Pull request merged**